### PR TITLE
Migrate treenode module.

### PR DESCRIPTION
**Pull request by flamingbear**
Migrate datatree's treenode module the base class representing node of a tree, with methods for accessing nodes and traversal.

- [X] completes step 2 `datatree/treenode.py` #8572
- [x] Tests added or updated
- [N/A] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [X] Internal Changes (including notable bug fixes) are documented in `whats-new.rst` 
- [N/A] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/tests/datatree/test_treenode.py
Comment (flamingbear) [inline xarray/tests/datatree/test_treenode.py:386]: @TomNicholas This test was passing, but probably incorrectly.  What should the `__str__` representation of a Named node look like?

From the test I was guessing:

``` 
NamedNode()
    NamedNode('Mary')
        NamedNode('Sam')
        NamedNode('Ben')
    NamedNode('Kate')
```

or as a string

``` 
"NamedNode()\n\tNamedNode('Mary')\n\t\tNamedNode('Sam')\n\t\tNamedNode('Ben')\n\tNamedNode('Kate')"
```

Should this be the `__str__` or the `__repr__`?
the `__repr__` is  currently:

```
"TreeNode(children={'Mary': TreeNode(children={'Sam': TreeNode(children={}), 'Ben': TreeNode(children={})}), 'Kate': TreeNode(children={})})"
```

File: xarray/tests/datatree/test_treenode.py
Comment (flamingbear) [inline xarray/tests/datatree/test_treenode.py:388]: failure here is expected right now.

File: xarray/tests/datatree/test_treenode.py
Comment (TomNicholas) [reply xarray/tests/datatree/test_treenode.py:392]: Ultimately both of these get overridden in `datatree.py` (`DataTree` is a subclass of `NamedNode`). So the printable representation of these internal classes should just be whatever is most convenient for us when debugging the code in the `treenode.py` module.

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:396]: I added a `__repr__` even if it's only ever going to be used in the test. Let me know if you hate it. 
7f2a178

File: doc/whats-new.rst
Comment (flamingbear) [inline doc/whats-new.rst:105]: @TomNicholas You want some acknowledgment that this is moving your code?  Maybe change Adds to migrates or include your name here?

Comment (TomNicholas) [reply doc/whats-new.rst:105]: I don't really need acknowledgement but I guess adding my name there might help someone later who wants to know who to blame for a bug they find? 😁

File: xarray/tests/datatree/test_treenode.py
Comment (TomNicholas) [reply xarray/tests/datatree/test_treenode.py:396]: > I added a __repr__ even if it's only ever going to be used in the test. Let me know if you hate it.

That seems fine. I would probably keep the quotation marks though because `Sam` is never an object, always a string.

> And I've found the render module and am less sure what this should have been testing.

Yeah... Do we even use the render module anymore?? The whole thing might just be a vestigial appendage that originated from the first prototype being based on the [`anytree` library](https://github.com/c0fec0de/anytree).

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:396]: > Do we even use the render module anymore?? 
At least some of it is used in the `render` function and both `__repr__` and `__str__` for a datatree.

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:395]: I'll call it add quotation marks. 548536b00

File: doc/whats-new.rst
Comment (flamingbear) [reply doc/whats-new.rst:105]: hopefully `git blame -C -C -C` will help out a lot but I'll update these.

File: xarray/tests/datatree/test_treenode.py
Comment (flamingbear) [inline xarray/tests/datatree/test_treenode.py:1]: One comment here: this file would slip into `xarray/tests/test_treenode.py` just as easily as into the datatree subdir.  Is there a preference?

I know at one point, the idea was that we would be able to review each PR as code was move up out of the `datatree_` subdirectory and if those PRs were well vetted we would be able to refactor the tests from `xarray/tests/datatree` into `xarray/tests` as made sense.  Is this still a plan with the new meetings to talk design issues?

**Comment by flamingbear** [comment]
> overall thoughts on the design / implementation here?

So my overall thoughts as I was moving it / grokking it was that the separation of concerns between the movement, linking of nodes and parenting rules vs the finding by names and paths as pretty clear.  The hooks for before and after attaching children seem thoughtful.  I haven't dived deep into the datatree code yet but this all makes sense so far.  If you were to scrap this for a nested dictionary, wouldn't you end up with pretty complicated code to handle moving/adding nodes? My gut says this is a solid choice to represent tree data.

File: xarray/tests/datatree/test_treenode.py
Comment (keewis) [reply xarray/tests/datatree/test_treenode.py:1]: did we discuss moving the `datatree` code into a submodule (like `xarray/core/datatree/`)? If we go that way, then I feel `xarray/tests/datatree` would make sense, otherwise it should be fine to have the tests in `xarray/tests`.

(I assume you meant moving from `xarray/datatree_/tests` to `xarray/tests`, right? If so, yes, I think that's still the plan, and we certainly can discuss this in the meetings)

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:1]: I don't know if there was a desire to move the entirety to a submodule e.g. `xarray/core/datatree`, but there was discussion about moving the tests from `datatree_/tests/x.py` to `xarray/tests/datatree/x.py` before eventually moving them to `xarray/tests/x.py` (or some matching test module).  I was wondering if that choice had changed with the new way forward.  that new way being, reviewing the decisions made in weekly meetings rather than for each PR.   That said, I don't have strong feelings or much history on how the structure of the repo was decided, so this was just a question of where this test file should land.

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:1]: and if it makes sense to move to a submodule, that makes a lot of these migrations even easier I think.

File: xarray/core/treenode.py
Comment (TomNicholas) [reply xarray/core/treenode.py:27]: Are we exposing `NodePath`? I suggested it, but after talking with @etienneschalk decided maybe it wasn't particularly helpful https://github.com/xarray-contrib/datatree/issues/205.

What's the logic for making it a pure path? The posix path is to ensure the slashes are in a consistent direction (remember this isn't a real filepath so shouldn't change automatically on windows systems).

Comment (keewis) [reply xarray/core/treenode.py:27]: I am not sure about this, but I believe the difference is that `PurePosixPath` doesn't have methods that require read/write access (like `touch`, `remove`, `mkdir`, `iterdir`, ...), while `PosixPath` does.

The main difference between filesystems and `datatree` is that there's only a single filesystem instance at a time, while you can have multiple `DataTree` objects at the same time. This means that a `NodePath` inheriting from `PosixPath` would have to be created from the `DataTree` object and would require a lot of discipline to not be confusing.

On the other hand, a `NodePath` that inherits from `PurePosixPath` would help a bit with path manipulations (`name`, `root`, `joinpath()`, `parts`, `parent`, `parents`, ...) that we would otherwise have to use `posixpath` for, but I didn't check if all of these exist in that module, and they might also not be as convenient to use.

Comment (TomNicholas) [reply xarray/core/treenode.py:27]: Sorry, I wrote my above comment too quickly - we both agree it should stay as a pure path.

We could also consider adding extra methods to this `NodePath` class if it simplifies other parts of the code.

File: xarray/tests/datatree/test_treenode.py
Comment (TomNicholas) [reply xarray/tests/datatree/test_treenode.py:1]: Does anyone have any strong opinions on this whatsoever? I don't, and if you don't, we should just do whatever is simplest for Matt.

Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:1]: I didn't have opinions except to see that other modules were well integrated and that you seemed to write the code to slip directly into the core.  I think some of the tests look like they will mesh well test_extensions for example.  I don't think it's more difficult per se.  I just don't have the previous decision history.   I personally, would probably rename test_io.py to something related to datatree io and move all of the tests into `xarray/tests`at the top level.  (I mean until something broke on me)

Comment (keewis) [reply xarray/tests/datatree/test_treenode.py:1]: let's discuss this in the meeting tomorrow, but yeah, I don't have any strong opinions.

**Comment by TomNicholas** [comment]
> personally, would probably rename test_io.py to something related to
datatree io and move all of the tests into xarray/testsat the top level.

That seems fine to me. Integrate the IO tests because the backends are
integrated. Separate the tests of tree functionality so that it's easy to
test different parts of the codebase independently during development
(that's why having a separate test_treenode.py is useful). But whether
test_treenode.py lives in a subdirectory or not just doesn't really matter
I think.

File: doc/whats-new.rst
Comment (flamingbear) [inline doc/whats-new.rst:141]: Is this allowed? (updating released changes)

Comment (TomNicholas) [reply doc/whats-new.rst:141]: If it's a correction / clarification then sure? I don't think we have a formal rule

Comment (keewis) [reply doc/whats-new.rst:141]: this should be fine, as long as it's just an update (for anything new I'd recommend a new entry).

File: xarray/core/treenode.py
Comment (etienneschalk) [reply xarray/core/treenode.py:27]: Related issue: https://github.com/xarray-contrib/datatree/issues/283 (Consistency between DataTree methods and pathlib.PurePath methods)

To me the graal would be to have datatree more and more compatible with `PurePosixPath` (what `NodePath` already does by inheriting from it, but we can imagine having an API accepting both strings and `PurePosixPath`, without the need to expose `NodePath` publicly).

As @keewis mentioned , there is a single instance of the filesystem whereas there can be multiple DataTrees. So a single DataTree could act like a "small filesystem on its own" and  implement the `PosixPath` methods. 

```
NodePath ~ PurePosixPath
DataTree ~ PosixPath 
``` 

The idea here is to delegate to `pathlib`  the diffcult task of choosing the best names for methods that help manipulating trees. For an audience of developers that are already used to writing scripts with `pathlib`, this would make working with datatree automatic! And this also gives plenty of ideas of features. Also, with hierarchical-formats such as Zarr that rely on the filesystem, openable by datatree, the proximity is innate.

Concrete example: working with multi-resolution rasters (a usecase of datatree). By using f-strings and glob, you can write resolution independant code once, like you would have done by manipulating a data structure on the file system

The schema on the [pathlib](https://docs.python.org/3/library/pathlib.html) doc (arrows are inheritance):

```mermaid
flowchart BT
  PurePosixPath --> PurePath
  Path --> PurePath
  PureWindowsPath --> PurePath
  PosixPath --> PurePosixPath
  PosixPath --> Path
  WindowsPath --> PureWindowsPath
  WindowsPath --> Path
```

How I see it:

```mermaid
flowchart BT
subgraph Pure
  PurePosixPath --> PurePath
  NodePath --> PurePosixPath
  end
subgraph Concrete
  DataTree --> NodePath
  PosixPath --> PurePosixPath
  end
```

(I removed `Path` for simplification)

I am not 100% sure it fully makes sense... The nuance is that when you instanciate a Path, it is always implicit that it is bound to the filesystem, while each DataTree instantiation creates its own data representation.

File: xarray/tests/datatree/test_treenode.py
Comment (flamingbear) [reply xarray/tests/datatree/test_treenode.py:1]: Will make this and the rest of the migrations flat.  I will start on moving the test_io.py in a separate PR off the tip of this one.

**Comment by flamingbear** [comment]
So that failure looks like a common one as opposed to something from this PR.  Is there anything specific to address in the PR before it can be merged?

**Comment by TomNicholas** [comment]
>  But whether `test_treenode.py` lives in a subdirectory or not just doesn't really matter
I think.

We made an executive decision in the [datatree meeting](https://github.com/pydata/xarray/issues/8747) today to not bother with subdirectories (for tests or code). We can always reorganise this later if we want to.

**Comment by TomNicholas** [comment]
> Is there anything specific to address in the PR before it can be merged?

I don't think so - @keewis what say you?

**Review by TomNicholas — APPROVED** [review]

File: xarray/core/treenode.py
Comment (TomNicholas) [reply xarray/core/treenode.py:27]: Thank you @etienneschalk for all your thoughts here!

> So a single DataTree could act like a "small filesystem on its own" and implement the PosixPath methods.
>
> I am not 100% sure it fully makes sense... The nuance is that when you instanciate a Path, it is always implicit that it is bound to the filesystem, while each DataTree instantiation creates its own data representation.

I strongly feel that DataTree / NodePath objects should not be associated with concrete filesystem paths.

However this

> To me the graal would be to have datatree more and more compatible with PurePosixPath (what NodePath already does by inheriting from it, but we can imagine having an API accepting both strings and PurePosixPath, without the need to expose NodePath publicly).

seems like a reasonable idea, and I think we should continue the discussion of that in https://github.com/xarray-contrib/datatree/issues/283.

For now I don't think there is anything we need to do in this PR specifically - we can revisit the API choices / accepting path-like types in a future PR.

**Review by shoyer — APPROVED** [review]

**Pull request merged**
